src/Stratis.Bitcoin.Features.Consensus/StakeValidator.cs

```diff
@@ -45,13 +48,137 @@ static StakeValidator()
+
+            // Genesis block.
+            if (chainedBlock == null)
+            {
+                clogger.LogTrace("(-)[GENESIS]:'{0}'", consensus.PowLimit);
+                return consensus.PowLimit;
+            }
+
+            // Find the last two blocks that correspond to the mining algo
+            // (i.e if this is a POS block we need to find the last two POS blocks).
+            BigInteger targetLimit = proofOfStake
+                ? consensus.ProofOfStakeLimitV2
+                : consensus.PowLimit.ToBigInteger();
+
+            // First block.
+            ChainedBlock pindexPrev = GetLastBlockIndex(stakeChain, chainedBlock, proofOfStake);
+            if (pindexPrev.Previous == null)
+            {
+                var res = new Target(targetLimit);
+                clogger.LogTrace("(-)[FIRST_BLOCK]:'{0}'", res);
+                return res;
+            }
+
+            // Second block.
+            ChainedBlock pindexPrevPrev = GetLastBlockIndex(stakeChain, pindexPrev.Previous, proofOfStake);
+            if (pindexPrevPrev.Previous == null)
+            {
+                var res = new Target(targetLimit);
+                clogger.LogTrace("(-)[SECOND_BLOCK]:'{0}'", res);
+                return res;
+            }
+
+            // This is used in tests to allow quickly mining blocks.
+            if (consensus.PowNoRetargeting)
+            {
+                clogger.LogTrace("(-)[NO_POW_RETARGET]:'{0}'", pindexPrev.Header.Bits);
+                return pindexPrev.Header.Bits;
+            }
+
+            int targetSpacing = TargetSpacingSeconds;
+            int actualSpacing = (int)(pindexPrev.Header.Time - pindexPrevPrev.Header.Time);
+            if (actualSpacing < 0)
+                actualSpacing = targetSpacing;
+
+            if (actualSpacing > targetSpacing * 10)
+                actualSpacing = targetSpacing * 10;
+
+            int targetTimespan = RetargetIntervalMinutes * 60;
+            int interval = targetTimespan / targetSpacing;
+
+            BigInteger target = pindexPrev.Header.Bits.ToBigInteger();
+
+            long multiplyBy = (interval - 1) * targetSpacing + actualSpacing + actualSpacing;
+            target = target.Multiply(BigInteger.ValueOf(multiplyBy));
+
+            long divideBy = (interval + 1) * targetSpacing;
+            target = target.Divide(BigInteger.ValueOf(divideBy));
+
+            clogger.LogTrace("The next target difficulty will be {0} times higher (easier to satisfy) than the previous target.", (double)multiplyBy / (double)divideBy);
+
+            if ((target.CompareTo(BigInteger.Zero) <= 0) || (target.CompareTo(targetLimit) >= 1))
+                target = targetLimit;
+
+            var finalTarget = new Target(target);
+            clogger.LogTrace("(-):'{0}'", finalTarget);
+            return finalTarget;
+        }
+
+        /// <summary>
+        /// Checks if provided <see cref="tx"/> is a valid coinstake transaction.
```

Comment:
Can you rename `tx` to transaction if you are already here?

Reply:
yeah i like the descriptions to use `see` only if necessary. If you use English it is often more readable, so if you write `Checks if the provided transaction is a valid coinstake transaction.` that's imho better

however, it seems that you are missing `param` comments. which are mandatory